java/flight/src/main/java/org/apache/arrow/flight/FlightStream.java

```diff
@@ -131,33 +131,53 @@ public DictionaryProvider takeDictionaryOwnership() {
     return provider;
   }
 
+  /**
+   * Get the descriptor for this stream. Only applicable on the server side of a DoPut operation. Will block until the
+   * client sends the descriptor.
+   */
   public FlightDescriptor getDescriptor() {
+    // This blocks until the schema message (with the descriptor) is sent.
+    getRoot();
     return descriptor;
   }
 
   /**
    * Closes the stream (freeing any existing resources).
    *
-   * <p>If the stream is isn't complete and is cancellable this method will cancel the stream first.</p>
+   * <p>If the stream is isn't complete and is cancellable, this method will cancel the stream first.</p>
    */
   public void close() throws Exception {
-    if (!completed && cancellable != null) {
-      cancel("Stream closed before end.", null);
+    RuntimeException cancelException = null;
+    try {
+      // cancel() can throw, but we still want to clean up resources.
+      if (!completed && cancellable != null) {
+        cancel("Stream closed before end.", null);
+      }
+    } catch (RuntimeException e) {
+      cancelException = e;
+    }
+
+    try {
+      List<AutoCloseable> closeables = ImmutableList.copyOf(queue.toArray()).stream()
+          .filter(t -> AutoCloseable.class.isAssignableFrom(t.getClass()))
+          .map(t -> ((AutoCloseable) t))
+          .collect(Collectors.toList());
+
+      final List<FieldVector> dictionaryVectors =
+          dictionaries == null ? Collections.emptyList() : dictionaries.getDictionaryIds().stream()
+              .map(id -> dictionaries.lookup(id).getVector()).collect(Collectors.toList());
+
+      // Must check for null since ImmutableList doesn't accept nulls
+      AutoCloseables.close(Iterables.concat(closeables,
+          dictionaryVectors,
+          applicationMetadata != null ? ImmutableList.of(root.get(), applicationMetadata)
```

Comment:
this might be clearer using ImmutableList.Builder outside of this method call instead of a ternary element.

Reply:
I see this isn't really something you've chagned though.

Reply:
I went ahead and cleaned it up since it was quite messy anyways.